### ISSUE TITLE
scripting-addition must be updated for macOS Sonoma 14.2

### ISSUE THREAD

**Issue by koekeishiya** [koekeishiya/yabai#2007]
title.

**Comment by roelvangils** [comment]
I am now using macOS 14.2 (23C64). I have updated the scripting addition and the `sudoers.d/yabai` file. The command `yabai -m query --windows` is functioning properly, but I am unable to switch spaces. Nothing happens. I have not encountered any error messages, including in the `yabai_username.err.log` file.

Should I temporarily revert to the default macOS method of switching spaces with animation, or is there something I am overlooking?

**Comment by airstrike** [comment]
> I am now using macOS 14.2 (23C64). I have updated the scripting addition and the `sudoers.d/yabai` file. The command `yabai -m query --windows` is functioning properly, but I am unable to switch spaces. Nothing happens. I have not encountered any error messages, including in the `yabai_username.err.log` file.
> 
> Should I temporarily revert to the default macOS method of switching spaces with animation, or is there something I am overlooking?

how did you update the scripting addition? I'm on MacOS 14.1.2 (23B92) and I'm getting

![image](https://github.com/koekeishiya/yabai/assets/152812/5cd910e2-4412-41f4-996d-967974f67a0c)

when I run 

```
yabai -m signal --add event=dock_did_restart action="sudo yabai --load-sa"
sudo yabai --load-sa
```

and I switch spaces with `yabai -m space --focus next`

**Comment by podocarp** [comment]
Same issue here. Things like toggling float on windows work, moving windows to spaces work, but switching to spaces (`yabai -m space --focus next` etc.) don't work for me.

**Comment by Khodaparastan** [comment]
same here

**Comment by SzymonMielecki** [comment]
same here but the issue seems similar to #1880

**Comment by alwyn** [comment]
@koekeishiya when you say "Scripting addition must be updated"  I am assuming it is a todo for you and not for the users?

**Comment by gldtn** [comment]
> @koekeishiya when you say "Scripting addition must be updated" I am assuming it is a todo for you and not for the users?

Yes, he mentioned somewhere that he will do that once it out of beta, I believe it's in RC(release candidate).

**Comment by cronJohn** [comment]
I encountered the same issue and have a workaround using MacOS's built-in shortcut for switching spaces/desktops.

1. Navigate to **System Preferences** > **Keyboard** > **Keyboard Shortcuts**.
2. Select **Mission Control** from the left sidebar.
3. Look for the option **Switch to Desktop ...** under the Mission Control dropdown.

I was concerned there would be a delay when switching spaces, which is why I used yabai, but that isn't the case. I should note that you will need to delete/comment out the `yabai -m space --focus` lines in your skhd config if that shortcut is the same as the one in the settings.

**Comment by kalebris** [comment]
yeah, that's the workaround I'm using but the workspace switches are slow, there is a significant delay compared to yabai workspace focus change. I'm personally using sketchybar on top, and that gets redrawn every time I do a switch.

**Comment by marsjane** [comment]
Same here, it is out of beta now, and after update, can not focus spaces, neither by click the sketchybar

**Comment by cronJohn** [comment]
@kalebris That could be because you have a shortcut set for moving left/right a space instead of switching to a specific desktop. There is a delay for me when moving left/right but not when switching to a specific desktop. If that doesn't work, it could be something with sketchybar.

**Comment by PrayagS** [comment]
macOS 14.2 23C64 is public now, no longer beta. Hope the fix is incoming soon 🤞🏼

**Comment by nibocn** [comment]
> I encountered the same issue and have a workaround using MacOS's built-in shortcut for switching spaces/desktops.
> 
> 1. Navigate to **System Preferences** > **Keyboard** > **Keyboard Shortcuts**.
> 2. Select **Mission Control** from the left sidebar.
> 3. Look for the option **Switch to Desktop ...** under the Mission Control dropdown.
> 
> I was concerned there would be a delay when switching spaces, which is why I used yabai, but that isn't the case. I should note that you will need to delete/comment out the `yabai -m space --focus` lines in your skhd config if that shortcut is the same as the one in the settings.

After I upgraded to macOS 14.2 (23C64), the `yabai -m space --focus` related commands could not be used normally. Even if I uninstalled yabai and reinstalled `scripting-addition`, it still did not work. After reading this method of reply by @cronJohn, I temporarily used this alternative method to achieve a similar effect.

First configure the Mission Control shortcut keys:
<img width="656" alt="image" src="https://github.com/koekeishiya/yabai/assets/10879287/74d8087e-6c75-48f4-8398-651b61aa3740">

Comment out the related shortcut key configuration for switching desktops in `skhdrc`:

<img width="499" alt="image" src="https://github.com/koekeishiya/yabai/assets/10879287/dacd2670-804a-492e-9dd0-9335656d3c09">

Then I also found that moving the application window to the specified desktop and then switching to the corresponding desktop would also cause problems. I also made corresponding adjustments to the configuration in `skhdrc`:

<img width="626" alt="image" src="https://github.com/koekeishiya/yabai/assets/10879287/d3120534-7580-4f23-be64-59d7a4571aca">

**Comment by steynh** [comment]
Oh nooo, I updated blindly, not realizing that this is a very new update and that yabai doesn't have the updated scripting addition yet. Thanks in advance for updating the scripting addition @koekeishiya, really appreciate your work! I can't work without yabai anymore xD

**Comment by TuralAsgar** [comment]
It doesn't work for me as well.
Script addition doesn't work.

I got so used to yabai, that without it I can't work.
Yabai is the reason why I am not complaining about macOS.

**Comment by 24unix** [comment]
> It doesn't work for me as well. Script addition doesn't work.
> 
> I got so used to yabai, that without it I can't work. Yabai is the reason why I am not complaining about macOS.

What is so important about Sonoma that you cannot wait a few days?
I'm still on Ventura, and will stay there until GPG Mail is fixed. Then I'll update, and be assured that Yabai is tested on Sonoma.

**Comment by roelvangils** [comment]
> What is so important about Sonoma that you cannot wait a few days? I'm still on Ventura, and will stay there until GPG Mail is fixed. Then I'll update, and be assured that Yabai is tested on Sonoma.

Sonoma is an exciting macOS update. Yabai worked well with the initial release and also with 14.1. I agree that, as Yabai "power users"", we should consider waiting for a week or two before upgrading to a point release. This will give the developer time to test and ensure Yabai works properly. It's always a cat and mouse game, unfortunately...

**Comment by rockofox** [comment]
The fork by @FelixKratz has been updated to 14.2 (https://github.com/FelixKratz/yabai/commit/df5b037108c4a70dc5e854bb60ccbff9701da4f5)

**Comment by schronck** [comment]
> The fork by @FelixKratz has been updated to 14.2 ([FelixKratz@df5b037](https://github.com/FelixKratz/yabai/commit/df5b037108c4a70dc5e854bb60ccbff9701da4f5))

Yeah but I still get the same error while trying to focus a space using his fork:
`cannot focus space due to an error with the scripting-addition.`

**Comment by kareemlukitomo** [comment]
> The fork by @FelixKratz has been updated to 14.2 ([FelixKratz@df5b037](https://github.com/FelixKratz/yabai/commit/df5b037108c4a70dc5e854bb60ccbff9701da4f5))

Thanks for this. I confirmed to successfully used his fork to correctly run `yabai --load-sa` on my 14.1.2 2022 M2 MBA

The steps working for me:
1. `sudo yabai --uninstall-sa && brew uninstall koekeishiya/formulae/yabai` uninstall the brew-version of yabai first
2. `git clone https://github.com/FelixKratz/yabai` cloned his fork
3. `cd yabai && make all` manually build the new/patched-yabai's binary
4. `cd bin && sudo mv yabai /usr/local/bin/yabai` moved this to $PATH (I think this is optional because somehow I need to manually specify the absolute path of the new yabai binary in my `.skhdrc`; up next) 
5. `nvim ~/.skhdrc` replace the "yabai" executable with absolute paths `/usr/local/bin/yabai`
6. Restart macOS
7. `sudo yabai --load-sa && yabai` run the new yabai. Maybe also try `brew services restart skhd`

Finally able to enjoy instantaneous space-switching again 🤭 hope this helps anyone. Hopefully later I able to help merge this diff to yabai's main branch.

**Comment by Jacky-Lzx** [comment]
> > The fork by @FelixKratz has been updated to 14.2 ([FelixKratz@df5b037](https://github.com/FelixKratz/yabai/commit/df5b037108c4a70dc5e854bb60ccbff9701da4f5))
> 
> Thanks for this. I confirmed to successfully used his fork to correctly run `yabai --load-sa` on my ~14.1.2~ macOS Sonoma 14.2 2022 M2 MBA. *edit: had just updated to 14.2 and still works
> 
> The steps working for me:
> 
> 1. `sudo yabai --uninstall-sa && brew uninstall koekeishiya/formulae/yabai` uninstall the brew-version of yabai first
> 2. `git clone https://github.com/FelixKratz/yabai` cloned his fork
> 3. `cd yabai && make all` manually build the new/patched-yabai's binary
> 4. `cd bin && sudo mv yabai /usr/local/bin/yabai` moved this to $PATH (I think this is optional because somehow I need to manually specify the absolute path of the new yabai binary in my `.skhdrc`; up next)
> 5. `nvim ~/.skhdrc` replace the "yabai" executable with absolute paths `/usr/local/bin/yabai`
> 6. Restart macOS
> 7. `sudo yabai --load-sa && yabai` run the new yabai. Maybe also try `brew services restart skhd`
> 
> Finally able to enjoy instantaneous space-switching again 🤭 hope this helps anyone. Hopefully later I able to help merge this diff to yabai's main branch. *edit: I meant only the scripting-addition fix, not Felix's personal patches of course.

This works for me. Previously I tried it but I failed. Now I successfully make it work on my M1 Pro Macbook with Sonoma 14.2.

Below are some notes that may help others:

1. You can just use the commit https://github.com/FelixKratz/yabai/commit/df5b037108c4a70dc5e854bb60ccbff9701da4f5 as a patch on yabai's latest commit (https://github.com/koekeishiya/yabai/commit/9fffe016ea91d221cb60891f1bb9b39fae738808).
2. You should restart your computer. (I think this is the reason I failed before)
3. Remember to configure the scripting addition as described [here](https://github.com/koekeishiya/yabai/wiki/Installing-yabai-(latest-release)#configure-scripting-addition).

**Comment by SzymonMielecki** [comment]
> > The fork by @FelixKratz has been updated to 14.2 ([FelixKratz@df5b037](https://github.com/FelixKratz/yabai/commit/df5b037108c4a70dc5e854bb60ccbff9701da4f5))
> 
> Thanks for this. I confirmed to successfully used his fork to correctly run `yabai --load-sa` on my ~14.1.2~ macOS Sonoma 14.2 2022 M2 MBA. *edit: had just updated to 14.2 and still works
> 
> The steps working for me:
> 
> 1. `sudo yabai --uninstall-sa && brew uninstall koekeishiya/formulae/yabai` uninstall the brew-version of yabai first
> 2. `git clone https://github.com/FelixKratz/yabai` cloned his fork
> 3. `cd yabai && make all` manually build the new/patched-yabai's binary
> 4. `cd bin && sudo mv yabai /usr/local/bin/yabai` moved this to $PATH (I think this is optional because somehow I need to manually specify the absolute path of the new yabai binary in my `.skhdrc`; up next)
> 5. `nvim ~/.skhdrc` replace the "yabai" executable with absolute paths `/usr/local/bin/yabai`
> 6. Restart macOS
> 7. `sudo yabai --load-sa && yabai` run the new yabai. Maybe also try `brew services restart skhd`
> 
> Finally able to enjoy instantaneous space-switching again 🤭 hope this helps anyone. Hopefully later I able to help merge this diff to yabai's main branch. *edit: I meant only the scripting-addition fix, not Felix's personal patches of course.

how do you detach it so that it can run in the background? 
`/usr/local/bin/yabai --start-service` does nothing

**Comment by alpiua** [comment]
@kareemlukitomo thanks, it works. But for some reason I can see a strange overlap on the second screen for 1 and 2 windows, and it overlap a sketchybar if there is only 1 window on a screen:

<img width="551" alt="image" src="https://github.com/koekeishiya/yabai/assets/14225967/b8069abc-a22c-42d9-91b0-b3e49c5fc915">

and join the @SzymonMielecki question - how to make this yabai version work as a service ?

**Comment by kareemlukitomo** [comment]
@SzymonMielecki @alpiua this is good enough for me (not exactly running as proper service though, since I don't have a good concept of how background service works on macOS)

https://chat.openai.com/share/348b1fd3-08c8-477a-8a3a-bdac028e6ec6

TL;DR

```shell
#!/bin/bash

while true; do
  yabai -c ~/.yabairc
  # Sleep for a bit to avoid rapid restarts in case of immediate failure
  sleep 1
done

```

Then running this into a detached screen/tmux/zellij session. (Gotta re-run on restart). Feel free to point out/give suggestions on how to improve this guys 👍

**Comment by kareemlukitomo** [comment]
For now, I'll just use Felix's personal brew formula ✌️ thank you for this @FelixKratz

1. `yabai --uninstall-sa && sudo rm -rf /usr/local/bin/yabai` cleanup first
2. `brew tap FelixKratz/homebrew-formulae` load this formula
3. `brew install --HEAD felixkratz/formulae/fyabai` install it
4. [Configure SA](https://github.com/koekeishiya/yabai/wiki/Installing-yabai-(latest-release)#configure-scripting-addition) and run yabai as usual 
5. `nvim ~/.skhdrc` revert the previous changes that I suggested above :b

Is there any benefit in cherry-picking https://github.com/FelixKratz/yabai/commit/df5b037108c4a70dc5e854bb60ccbff9701da4f5 and creating a PR for the master branch in this original repo? Happy to help with this or anything else needed. Cheers!

**Comment by rwijtvliet** [comment]
This seems to fit here, but let me know if I should start a new thread instead:

I'm new to mac, and it seems these OS updates are a lot more common than I expected. Is there a way to tell if an OS update will break yabai, before updating? My company aggressively pushes OS updates, and while I can postpone them for a bit, it's cumbersome.

**Comment by PrayagS** [comment]
Was able to cherry-pick the commit from Felix's fork (https://github.com/PrayagS/yabai/tree/patched-for-14.2) and build the updated binary.

Tried placing the binary at yabai's original location (`/opt/homebrew/bin/yabai`) hoping that the background service commands would work but they didn't 🥲.

Used this old hack that I used to have when `skhd` didn't have a background service,
```sh
# Start skhd/yabai
if pgrep -x "yabai" >/dev/null
then
else
   yabai
   echo "yabai started"
fi
```

Placed this in my `.zshenv` and yabai is running as long as there's a terminal window open.

**Comment by encinascarles** [comment]
> For now, I'll just use Felix's personal brew formula ✌️ thank you for this @FelixKratz
> 
> 1. `yabai --uninstall-sa && sudo rm -rf /usr/local/bin/yabai` cleanup first
> 2. `brew tap FelixKratz/homebrew-formulae` load this formula
> 3. `brew install --HEAD felixkratz/formulae/fyabai` install it
> 4. [Configure SA](https://github.com/koekeishiya/yabai/wiki/Installing-yabai-(latest-release)#configure-scripting-addition) and run yabai as usual
> 5. `nvim ~/.skhdrc` revert the previous changes that I suggested above :b
> 
> Is there any benefit in cherry-picking [FelixKratz@df5b037](https://github.com/FelixKratz/yabai/commit/df5b037108c4a70dc5e854bb60ccbff9701da4f5) and creating a PR for the master branch in this original repo? Happy to help with this or anything else needed. Cheers!

It worked for me, thank you so much!

**Comment by boronine** [comment]
I can confirm v6.0.2 works!

Just by eliminating the workspace switching animation, yabai is saving an hour of my life every year, to say nothing of the emotional damage that vanilla macOS window management causes. Yabai is easily providing $100+ of value per year.

Åsmund's work finally inspired me to set up a Patreon account.

Thank you Åsmund! 🙏 🙏 🙏 

-----> https://www.patreon.com/aasvi

**Comment by marsjane** [comment]
Merry Christmas!! However I updated yabai to v6.0.2, but it seems still not working:
```
~ ❯ yabai -m space --focus 2
cannot focus space due to an error with the scripting-addition.
~ ❯ yabai --version
yabai-v6.0.2
```
anything I missed here? I tried to run sudo yabai --load-sa manually, there is no error, just not woking as expected

**Comment by boronine** [comment]
@marsjane did you edit your sudoers file after upgrading? In particular, the sha256 must be updated for the new binary

https://github.com/koekeishiya/yabai/wiki/Installing-yabai-(latest-release)#configure-scripting-addition

**Comment by marsjane** [comment]
Oh yes it works! Sorry for missing this, this is the first time I update yabai ;) Many thanks !! @boronine

**Comment by TuralAsgar** [comment]
It works!
Thank you!

**Comment by Tchoupinax** [comment]
Hey,
With Sonoma 14.2.1, it didn't work. I spent 1 hour to try to make it working again but without any success. The patch and the fork of @FelixKratz worked! Thank you a lot!

**Comment by d34dh0r53** [comment]
> Was able to cherry-pick the commit from Felix's fork (https://github.com/PrayagS/yabai/tree/patched-for-14.2) and build the updated binary.
> 
> Tried placing the binary at yabai's original location (`/opt/homebrew/bin/yabai`) hoping that the background service commands would work but they didn't 🥲.
> 
> Used this old hack that I used to have when `skhd` didn't have a background service,
> 
> ```shell
> # Start skhd/yabai
> if pgrep -x "yabai" >/dev/null
> then
> else
>    yabai
>    echo "yabai started"
> fi
> ```
> 
> Placed this in my `.zshenv` and yabai is running as long as there's a terminal window open.

The `disown` command will let you detach the `yabai` process from the controlling TTY allowing you to close the terminal window that you launched `yabai` from.  You can run something like:

`yabai -V >> /tmp/yabai-HEAD.log 2>&1 &` to launch the daemon and then `disown` will detach the process from your terminal and allow you to close it.  Hope this helps :)

**Comment by williamhCode** [comment]
I'm trying to get the scripting addition to work with Yabai on macos 14.2.1 with no luck. I setup everything and updated the sudoers file, but it gives this error when i try to run yabai:
```
yabai: scripting-addition failed to inject payload into Dock.app!
```

**Comment by nagy135** [comment]
> For now, I'll just use Felix's personal brew formula ✌️ thank you for this @FelixKratz
> 
> 1. `yabai --uninstall-sa && sudo rm -rf /usr/local/bin/yabai` cleanup first
> 2. `brew tap FelixKratz/homebrew-formulae` load this formula
> 3. `brew install --HEAD felixkratz/formulae/fyabai` install it
> 4. [Configure SA](https://github.com/koekeishiya/yabai/wiki/Installing-yabai-(latest-release)#configure-scripting-addition) and run yabai as usual
> 5. `nvim ~/.skhdrc` revert the previous changes that I suggested above :b
> 
> Is there any benefit in cherry-picking [FelixKratz@df5b037](https://github.com/FelixKratz/yabai/commit/df5b037108c4a70dc5e854bb60ccbff9701da4f5) and creating a PR for the master branch in this original repo? Happy to help with this or anything else needed. Cheers!

This didnt work for me since it just installed different version of yabai and didnt link it to system. So I added `brew uninstall yabai` and `brew link fyabai`.

**Comment by hkgnp** [comment]
I would like to confirm that scripting addition on 6.0.6 works.

**Comment by leohxj** [comment]
> I would like to confirm that scripting addition on 6.0.6 works.
no

**Comment by weeebdev** [comment]
@koekeishiya is yabai managed by homebrew services or plist for now?
